Add workspace root content type

raphael-s opened the pull request:
Implements a new content type "WorkspaceRoot" which is the folderish parent for workspaces. (Workspace content type will be added later)

closes #3675

phgross commented on opengever/core/profiles/default/types/opengever.workspace.root.xml:
```diff
@@ -0,0 +1,68 @@
+  <!-- enabled behaviors -->
+  <property name="behaviors">
+    <element value="opengever.tabbedview.interfaces.ITabbedViewEnabled" />
+    <element value="opengever.sharing.behaviors.IDossier" />
```
Is that really necessary, as soon as I know the workspace part should habe their own roles and permissions.

raphael-s replied:
I used the xml of `opengever.repository.repositoryroot` as a base for this. I removed the behaviour now.

phgross commented on opengever/core/profiles/default/types/opengever.workspace.root.xml:
```diff
@@ -0,0 +1,68 @@
+  <alias from="view" to="@@view" />
+
+  <!-- Actions -->
+
```
✂️

phgross commented on opengever/core/upgrades/20171206170115_add_workspace_root_content_type/upgrade.py:
```diff
@@ -0,0 +1,13 @@
+
+    def __call__(self):
+        self.install_upgrade_profile()
+        self.update_workflow_security(
```
what ??? 😞

raphael-s replied:
Clicked the button in lawgiver, I thought this would be necessary since I added a new permission. Seems like it is actually not needed, I removed the workflow update completely.

phgross commented on opengever/core/upgrades/20171206170115_add_workspace_root_content_type/workflows/opengever_committee_workflow/definition.xml:
```diff
@@ -0,0 +1,716 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
??? 😞 

--> 🗑

phgross commented on opengever/core/upgrades/20171206170115_add_workspace_root_content_type/workflows/opengever_committeecontainer_workflow/definition.xml:
```diff
@@ -0,0 +1,401 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
--> 🗑

phgross commented on opengever/setup/profiles/default_content/content_creation/01_default_content.json:
```diff
@@ -24,5 +24,12 @@
+        "_path": "teamraeume",
+        "_type": "opengever.workspace.root",
+        "title_de": "Teamräume",
+        "title_fr": "Teamräume",
```
ask niklaus for a meaningful translation.

phgross commented on opengever/setup/profiles/default_content/content_creation/01_default_content.json:
```diff
@@ -24,5 +24,12 @@
+        "_type": "opengever.workspace.root",
+        "title_de": "Teamräume",
+        "title_fr": "Teamräume",
+        "_transitions": "publish"
```
is there a publish transition ...

raphael-s replied:
Yes, there is. Have a look at the item above.

It works, just tested it. The item is published after its creation.

phgross commented on opengever/testing/fixtures.py:
```diff
@@ -805,6 +806,11 @@ def create_committee(self, title, repository_folder, group_id,
+    def create_workspaces(self):
+        self.workspace_root = self.register('workspace_root', create(
+            Builder('workspace_root').having(title_de=u'Teamr\xe4ume',
+                                             title_fr=u'Teamr\xe4ume')))
```
same here

phgross commented on opengever/workspace/content/configure.zcml:
```diff
@@ -0,0 +1,6 @@
+<configure
```
We do not have a `content` subdirectory in GEVER, i would move all that to opengever/workspace

jone commented on opengever/core/profiles/default/types/opengever.workspace.root.xml:
```diff
@@ -0,0 +1,68 @@
+  <property name="description" i18n:translate="" />
+  <property name="icon_expr" />
+  <property name="allow_discussion">False</property>
+  <property name="global_allow">True</property>
```
-> `False`: you've already added it to the allowed-types on `Plone Site`.

jone commented on opengever/core/profiles/default/types/opengever.workspace.root.xml:
```diff
@@ -0,0 +1,68 @@
+  <property name="default_view">tabbed_view</property>
+  <property name="default_view_fallback">False</property>
+  <property name="view_methods">
+    <element value="view" />
```
✂️ I think we only want a tabbedview here.
Having two values here probably lets the layout-menu pop up; we don't want that in GEVER.

jone commented on opengever/testing/fixtures.py:
```diff
@@ -805,6 +806,11 @@ def create_committee(self, title, repository_folder, group_id,
                     group_id=group_id))
         return committee
 
+    def create_workspaces(self):
```
Please rename this to `create_workspace_root`.

Creating the workspaces should be separated and executed by another user in another freezed hour.

jone commented on opengever/workspace/browser/workspace_root.py:
```diff
@@ -0,0 +1,22 @@
+
+
+@adapter(IFolderish, IDefaultBrowserLayer, IDexterityFTI)
+class AddView(DefaultAddView):
```
This view is not registered.
How do we test that?

jone commented on opengever/workspace/content/workspace_root.py:
```diff
@@ -0,0 +1,18 @@
+        need any specific schema fields """
+
+
+alsoProvides(IWorkspaceRootSchema, IFormFieldProvider)
```
In GEVER we prefer to use the decorator style.
So please use
```python
@provider(IFormFieldProvider)
class IWorkspaceRootSchema(model.Schema):
    …
```

See https://github.com/4teamwork/opengever.core/blob/master/opengever/meeting/proposaltemplate.py#L14
I'm aware that there are places where this is not yet converted.

raphael-s commented:
@jone @phgross I implemented your requested changes. Is it okay now?

jone commented:
Der Titel fehlt.
![bildschirmfoto 2017-12-11 um 13 36 04](https://user-images.githubusercontent.com/7469/33831416-59a6bff0-de78-11e7-9d7c-ae6dd0392cfb.png)

jone commented:
Bitte Standard-Workflow entfernen: der Typ soll momentan keinen Workflow haben, das kommt später.
![bildschirmfoto 2017-12-11 um 13 37 48](https://user-images.githubusercontent.com/7469/33831508-9952b334-de78-11e7-8692-74256019ea7f.png)

phgross commented on opengever/workspace/workspace_root.py:
```diff
@@ -0,0 +1,16 @@
+        need any specific schema fields """
+
+
+class WorkspaceRoot(Item):
```
Funktioniert die TranslatedTitle Funktionalität wirklich wie gewünscht? Aus meiner Sicht müsste hier das `TranslatedTitleMixin` verwendet werden. 🤔

raphael-s commented:
@jone I removed the default workflow and added the TranslatedTitleMixin the way @phgross suggested it, the title is now displayed correctly. i18n translations for the FTI are now available aswell.

jone commented on opengever/core/locales/de/LC_MESSAGES/opengever.core.po:
```diff
@@ -231,6 +231,10 @@ msgstr "Amtswechsler umschalten"
 
+#: ./opengever/core/profiles/default/types/opengever.workspace.root.xml
+msgid "WorkspaceRoot"
+msgstr "Teamruäume"
```
typo

jone commented on opengever/setup/profiles/default_content/content_creation/01_default_content.json:
```diff
@@ -24,5 +24,12 @@
         "_transitions": "publish"
+    },
+    {
+        "_path": "teamraeume",
```
Ich finde die ID nicht schön. Wie wärs mit `workspaces`?

jone commented on opengever/testing/fixtures.py:
```diff
@@ -850,6 +851,11 @@ def create_committee(self, title, repository_folder, group_id,
+    def create_workspace_root(self):
+        self.workspace_root = self.register('workspace_root', create(
+            Builder('workspace_root').having(title_de=u'Teamr\xe4ume',
+                                             title_fr=u'Espace partag\xe9')))
```
Kann man hier die ID setzen? Wäre cool wenn wir die gleiche ID hätten wie im Example Content.

raphael-s commented:
@jone Done 👍

jone commented on opengever/workspace/workspace_root.py:
```diff
@@ -0,0 +1,19 @@
+class WorkspaceRoot(Item, TranslatedTitleMixin):
+    implements(IWorkspaceRoot)
+
+    Title = TranslatedTitleMixin.Title
```
@raphael-s mich hat schon vorher irritiert wie die Reihenfolge der vererbten Klasse ist; diese Zeile erklärts nun 😉 

Ich schlage vor, dass du diese Zeile entfernst und dafür die Reihenfolge der Vererbten Superklassen umkehrst:
```diff
-class WorkspaceRoot(Item, TranslatedTitleMixin):
+class WorkspaceRoot(TranslatedTitleMixin, Item):
```

Die MRO funktioniert so, dass die Klassen links die Klassen rechts überschreiben.

Ich würde dies machen auch wenn die anderen Orte im GEVER nicht so sind. Weil das Umkehren der MRO ist eigentlich korrekt und es hat keine Seiteneffekte.

raphael-s replied:
I changed the order as you requested, now the tests fail. Also the title is no longer translated. Are you sure that the override order is left to right?

raphael-s replied:
Okay, I think I got it now 😄 

But the tests are still failing... what am I doing wrong?

raphael-s replied:
@jone The inheritance order has to be the way it was before because `Item` does some magic which gets broken by changing the order. See -> https://github.com/4teamwork/opengever.core/pull/1394
